### PR TITLE
Git_Basics.rst : Put the commands from http://ohshitgit.com/

### PR DESCRIPTION
**Pull request by sourojit**
Some important git commands from http://ohshitgit.com/
These are helpful when something goes wrong.

Closes https://github.com/coala/coala/issues/4985

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by shreyateeza** [comment]
@sourojit You need to squash your commits. You need to write commit in imperative present tense. Also read this https://coala.io/commits.

**Comment by sourojit** [comment]
@shreyateeza after squashing the commits somehow the commit message for coalibs

**Comment by sourojit** [comment]
@shreyateeza Fixed the problem.Thanks . 👍

**Comment by sourojit** [comment]
@shreyateeza Please review this PR

File: docs/Developers/Git_Basics.rst [hunk 1]
```diff
@@ -542,3 +542,67 @@ also reorder the commit listing to change the order of the commits themselves.
 If you would like more information/commands, please use your favourite search
```
Comment (vddesai1998) [inline docs/Developers/Git_Basics.rst]: Add this paragraph at bottom of the file

File: docs/Developers/Git_Basics.rst [hunk 2]
```diff
@@ -542,3 +542,67 @@ also reorder the commit listing to change the order of the commits themselves.
...
+on the terminal and each command will have an index. The structure of the
+index is HEAD@{index}.This command is very important when you are in a mess.
+The next command will provide the users a weapon to fix their mistakes.
+
+
+::
+
+   $ git reset HEAD@{index}
+
+Reset to the point before you started messing up. HEAD@{index} denotes the
```
Comment (vddesai1998) [inline docs/Developers/Git_Basics.rst]: Enclose " HEAD@{index} " with two backquotes ( ``  ) like you have done at other places

**Comment by sourojit** [comment]
@vddesai1998 please review this PR

**Comment by vddesai1998** [comment]
@sourojit  Now you have to squash the commits into one.

**Comment by sourojit** [comment]
@vddesai1998 done 👍

**Comment by sourojit** [comment]
@vddesai1998 now what should be done??

**Comment by sourojit** [comment]
@shreyateeza @vddesai1998 what to do next?? I am not familiar with the process.

**Comment by shreyateeza** [comment]
@sourojit you need to have patience till it get merged. Feel free to take any other issue and work on it. :+1:

**Comment by sourojit** [comment]
Ok thanks 😀

**Comment by troublemagnet** [comment]
Please refer to http://api.coala.io/en/latest/Developers/Writing_Good_Commits.html and change your commit message. Apart from that, please work on the gitmate-bot's messages.

**Comment by sourojit** [comment]
@troublemagnet now the gitmate-bot suggests that all tests have been successfully passed

**Comment by Makman2** [comment]
Please add the issue reference also to the commit message, not only the PR.

**Comment by Makman2** [comment]
unack 5fd0cda

**Comment by sourojit** [comment]
@Makman2 should I append "closes #4985" to the already existing commit message

**Comment by Makman2** [comment]
> should I append "closes #4985" to the already existing commit message

yes, but please use the full link + "closes" with upper-case-C ;) --> "Closes"

**Comment by Makman2** [comment]
Btw, are committing with a different account?

**Comment by Makman2** [comment]
Just curious, any reason for that? :)

**Comment by sourojit** [comment]
@Makman2 changing the commit message causes the gitmate-bot to generate the above test failed message. It says the no. of characters in the commit message is more than allowed.

**Comment by sourojit** [comment]
The commits occur through a different account even though I commit through this account. There is no reason to this unusual behavior and I plan on fixing it soon. :D

**Comment by Makman2** [comment]
Yes because the issue reference also appears in the shortlog, please put it into the commit body :+1: (see other commits in coala)

**Comment by sourojit** [comment]
@Makman2 done 👍

File: docs/Developers/Git_Basics.rst [hunk 3]
```diff
@@ -539,6 +539,69 @@ commands (described below) for each commit to be rebased. These commands
...
+   $ git reset HEAD~ --hard
+   $ git checkout new-branch-name
+
+These commands are useful if you have accidently commited something to master
+but should have been commited to a new branch. ``git branch new-branch-name``
+creates a new branch from the current state of the master.
+``git reset HEAD~ --hard`` removes the commit from the master branch.
+After running ``git checkout new-branch-name`` commit into this branch.
+Note:this does not work if you have already pushed to origin or run other
+commands before trying this.Then you have to use ``git reset HEAD@{index}``.
```
Comment (stobis) [inline docs/Developers/Git_Basics.rst]: Missing space before `Then` (new phrase beginning)

File: docs/Developers/Git_Basics.rst [hunk 4]
```diff
@@ -539,6 +539,69 @@ commands (described below) for each commit to be rebased. These commands
...
+   $ git branch new-branch-name
+   $ git reset HEAD~ --hard
+   $ git checkout new-branch-name
+
+These commands are useful if you have accidently commited something to master
+but should have been commited to a new branch. ``git branch new-branch-name``
+creates a new branch from the current state of the master.
+``git reset HEAD~ --hard`` removes the commit from the master branch.
+After running ``git checkout new-branch-name`` commit into this branch.
+Note:this does not work if you have already pushed to origin or run other
```
Comment (stobis) [inline docs/Developers/Git_Basics.rst]: Space after `:`?

File: docs/Developers/Git_Basics.rst [hunk 5]
```diff
@@ -539,6 +539,69 @@ commands (described below) for each commit to be rebased. These commands
...
+
+   $ git reset HEAD~ --soft
+   $ git stash
+   $ git checkout correct-branch
+   $ git stash pop
+   $ git add <file/folder_name>
+   $ git commit -m "your message"
+
+These series of commands are used when you have accidentally commited to the
+wrong branch.``git reset HEAD~ --soft`` undos the last commit and
```
Comment (stobis) [inline docs/Developers/Git_Basics.rst]: screen shot 2018-01-14 at 23 18 28
Just asking: Is it possible to do something with that?

File: docs/Developers/Git_Basics.rst [hunk 6]
```diff
@@ -539,6 +539,69 @@ commands (described below) for each commit to be rebased. These commands
...
+``git stash pop`` applies the commit and removes it from the stack.
+Then using ``git add <file/folder_name>`` and the commit command
+``git commit -m "your message"`` commit your changes to the correct branch.
+
+
+::
+
+   $ git diff --staged
+
+``git diff`` Show changes between the working tree and the index or a tree,
```
Comment (stobis) [inline docs/Developers/Git_Basics.rst]: Non-capital letter in `Show` to keep convention

**Comment by Makman2** [comment]
Actually I just thought of putting the link somewhere, I fear a bit we could violate copyrights here somehow :/

**Comment by Makman2** [comment]
unack 20fbf8bb17c82e6031ceae2bcaaae6227557d62c

**Comment by sourojit** [comment]
@Makman2 so should I just include the link in GitHub.rst file or should It be included somewhere else

**Comment by Makman2** [comment]
> so should I just include the link in GitHub.rst file or should It be included somewhere else

Yeah just include the link inside GitHub.rst :+1:

**Comment by sourojit** [comment]
@Makman2 done 👍

File: docs/Developers/Git_Basics.rst [hunk 7]
```diff
@@ -539,6 +539,8 @@ commands (described below) for each commit to be rebased. These commands
 determine how individual commits will be transferred to the new base. You can
 also reorder the commit listing to change the order of the commits themselves.
 
+For more commands go to http://ohshitgit.com/
```
Comment (Makman2) [inline docs/Developers/Git_Basics.rst]: 1. Missing period
2. This doesn't read so nice and is not really helpful, it's just a dropped link. You should try to highlight the fact that this link is very useful if something goes wrong. It's not just a bunch of commands^^ (well in fact it is, but commands helping you to get out of a screwed-state ;) )

**Comment by sourojit** [comment]
@Makman2 what is Travis CI build failed and circleCI build failed? I couldn't quite understand the information provided in the details link of the build check. The only conclusion I  could draw is that Travis CI  is related to python versions and couldn't make any head or tail out of circleCI.

**Comment by Makman2** [comment]
Closing due to inactivity.

**Pull request closed without merging**